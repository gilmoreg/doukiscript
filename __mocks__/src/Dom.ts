import { IDomMethods } from "../../src/Dom";

export default class FakeDomMethods implements IDomMethods {
    dateSetting: string
    constructor(dateSetting = 'a') {
        this.dateSetting = dateSetting;
    }

    addDropDownItem() { }

    addImportForm(syncFn: Function) {
        return jest.fn();
    }

    getDateSetting(): string {
        return this.dateSetting;
    };

    getDebugSetting(): boolean {
        return false;
    }

    getCSRFToken(): string {
        return 'csrfToken';
    }

    getMALUsername(): string {
        return 'malUsername';
    }

    getAnilistUsername(): string {
        return 'anilistUsername';
    }
}