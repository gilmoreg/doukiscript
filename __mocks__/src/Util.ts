const mockUtil = {
    sleep: () => Promise.resolve(),
    id: (str: string) => `#${str}`,
    getOperationDisplayName: (operation: string) => operation,
}

export default mockUtil;