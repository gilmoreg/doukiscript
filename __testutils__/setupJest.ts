import * as fetch from 'node-fetch';
// @ts-ignore
global.fetch = fetch;

class FakeDOM {
    querySelector(selector: string) {
        switch (selector) {
            case 'add_anime_comments':
            case 'add_manga_comments':
                return {
                    value: 'comments'
                }
            default:
                return {
                    value: '0'
                }
        }
    }
}

class FakeXHR {
    onload: Function = () => { };
    onerror: Function = () => { };
    responseType: string = 'document';
    open(method: string, url: string) {
        console.log(`Making XHR ${method} request to ${url}`);
    }
    send() {
        this.onload();
    }
    responseXML: FakeDOM

    constructor() {
        this.responseXML = new FakeDOM();
    }
}
// @ts-ignore
global.XMLHttpRequest = FakeXHR;