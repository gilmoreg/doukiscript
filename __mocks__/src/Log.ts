import { ILog } from "../../src/Log";

export default class FakeLog implements ILog {
    clear() {
        console.log('clearing console');
    }
    error(msg: string) {
        console.log('logging error', msg);
    }
    info(msg: string) {
        console.log('logging info', msg);
    }
    addCountLog(op: string, type: string, max: number) {
        console.log('creating count log for', op, type, max);
    }
    updateCountLog(op: string, type: string, ct: number) {
        console.log('updating count log for', op, type, ct);
    }
}





