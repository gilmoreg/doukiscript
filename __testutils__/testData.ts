import * as Types from '../src/Types';
import { IDomMethods } from '../src/Dom';
import { DIContainer } from '../src/DIContainer';
import FakeLog from '../__mocks__/Log';
import { FakeMALForm } from '../__mocks__/MALForm';

const createDate = (year = 0, month = 0, day = 0) => ({ year, month, day });

const malItem: Types.BaseMALItem = {
    status: 2,
    csrf_token: 'csrfToken',
    score: 10,
    finish_date_string: null,
    start_date_string: null,
};

const malAnime = {
    ...malItem,
    anime_id: 1,
    num_watched_times: 1,
    num_watched_episodes: 12,
    anime_airing_status: 2,
    anime_num_episodes: 12,
} as Types.MALLoadAnime;

const malManga = {
    ...malItem,
    manga_id: 2,
    num_read_chapters: 12,
    num_read_times: 1,
    num_read_volumes: 1
} as Types.MALLoadManga;

export const createFakeMALAnime = (data: any = {}) =>
    ({ ...malAnime, ...data });
export const createFakeMALManga = (data: any = {}) =>
    ({ ...malManga, ...data });

const alAnime: Types.FormattedEntry = {
    status: 'COMPLETED',
    score: 10,
    progress: 12,
    progressVolumes: 0,
    startedAt: createDate(),
    completedAt: createDate(),
    repeat: 1,
    id: malAnime.anime_id,
    title: 'title',
    type: 'anime'
};

const alManga: Types.FormattedEntry = {
    status: 'COMPLETED',
    score: 10,
    progress: 12,
    progressVolumes: 1,
    startedAt: createDate(),
    completedAt: createDate(),
    repeat: 1,
    id: malManga.manga_id,
    title: 'title',
    type: 'manga'
};

export const createFakeAnilistAnime = (data: any = {}): Types.FormattedEntry =>
    ({ ...alAnime, ...data });
export const createFakeAnilistManga = (data: any = {}): Types.FormattedEntry =>
    ({ ...alManga, ...data });

export const createFakeDomMethods = (dateSetting = 'a'): IDomMethods => ({
    addDropDownItem: jest.fn(),
    addImportForm: (syncFn: Function) => jest.fn(),
    getDateSetting: jest.fn().mockImplementation().mockReturnValue(dateSetting),
    getCSRFToken: () => 'csrfToken',
    getMALUsername: () => 'malUsername',
    getAnilistUsername: () => 'anilistUsername'
});

export const defaultMocks = () => new DIContainer(undefined, () => new FakeMALForm(), createFakeDomMethods(), new FakeLog());