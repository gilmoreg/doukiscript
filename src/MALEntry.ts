import * as Dom from './Dom';
import * as T from "./Types";

export const createMALEntry = (al: T.FormattedEntry, mal: T.MALLoadItem, csrfToken: string) =>
    al.type === 'anime' ? new MALEntryAnime(al, mal, csrfToken) : new MALEntryManga(al, mal, csrfToken);

const MALStatus = {
    Current: 1,
    Completed: 2,
    Paused: 3,
    Dropped: 4,
    Planning: 6
}

const getStatus = (status: string) => {
    // MAL status: 1/watching, 2/completed, 3/onhold, 4/dropped, 6/plantowatch
    // MAL handles REPEATING as a boolean, and keeps status as COMPLETE
    switch (status.trim()) {
        case 'CURRENT':
            return MALStatus.Current;
        case 'REPEATING':
        case 'COMPLETED':
            return MALStatus.Completed;
        case 'PAUSED':
            return MALStatus.Paused;
        case 'DROPPED':
            return MALStatus.Dropped;
        case 'PLANNING':
            return MALStatus.Planning;
        default:
            throw new Error(`unknown status "${status}"`);
    }
}

const buildDateString = (date: T.MediaDate) => {
    if (date.month === 0 && date.day === 0 && date.year === 0) return null;
    const dateSetting = Dom.getDateSetting();
    const month = `${String(date.month).length < 2 ? '0' : ''}${date.month}`;
    const day = `${String(date.day).length < 2 ? '0' : ''}${date.day}`;
    const year = `${date.year ? String(date.year).slice(-2) : 0}`;
    if (dateSetting === 'a') {
        return `${month}-${day}-${year}`;
    }
    return `${day}-${month}-${year}`;
}

const createMALFormData = (malData: T.MALFormData): string => {
    let formData = '';
    Object.keys(malData).forEach(key => {
        formData += `${encodeURIComponent(key)}=${encodeURIComponent(malData[key])}&`;
    });
    return formData.replace(/&$/, '');
}

export interface MALEntry {
    shouldUpdate(): boolean
    shouldAdd(): boolean
    formData(): string
    type: string
    id: number
    title: string
    postData: T.MALPostItem
}

export class BaseMALEntry implements MALEntry {
    alData: T.FormattedEntry
    malData: T.MALLoadItem
    _postData: T.MALPostItem
    csrfToken: string

    constructor(al: T.FormattedEntry, mal: T.MALLoadItem, csrfToken: string = '') {
        this.alData = al;
        this.malData = mal;
        this.csrfToken = csrfToken;
        this._postData = this.createPostData();
    }

    protected createBaseMALPostItem(): T.MALPostItem {
        return {
            status: getStatus(this.alData.status),
            csrf_token: this.csrfToken,
            score: this.alData.score || 0,
            finish_date: {
                year: this.alData.completedAt.year || 0,
                month: this.alData.completedAt.month || 0,
                day: this.alData.completedAt.day || 0
            },
            start_date: {
                year: this.alData.startedAt.year || 0,
                month: this.alData.startedAt.month || 0,
                day: this.alData.startedAt.day || 0
            }
        } as T.MALPostItem;
    }

    shouldUpdate(): boolean {
        return Object.keys(this._postData).some(key => {
            switch (key) {
                case 'csrf_token':
                case 'anime_id':
                case 'manga_id':
                    return false;
                case 'start_date':
                case 'finish_date':
                    {
                        // @ts-ignore
                        const dateString = buildDateString(this._postData[key]);
                        if (dateString !== this.malData[`${key}_string`]) {
                            return true;
                        }
                        return false;
                    }
                case 'num_read_chapters':
                case 'num_read_volumes':
                case 'num_watched_episodes':
                    // Anlist and MAL have different volume, episode, and chapter counts for some media;
                    // If the item is marked as completed, ignore differences (Status 2 is COMPLETED)
                    // EXCEPT when the count is 0, in which case this was newly added without a count and needs
                    // to be updated now that the count is available
                    {
                        if (this.malData.status === MALStatus.Completed && this.malData[key] !== 0) {
                            return false;
                        }
                        if (this._postData[key] !== this.malData[key]) {
                            return true;
                        };
                        return false;
                    }
                // In certain cases the next two values will be missing from the MAL data and trying to update them will do nothing.
                // To avoid a meaningless update every time, skip it if undefined on MAL
                case 'num_watched_times':
                case 'num_read_times':
                    {
                        if (!this.malData.hasOwnProperty(key)) {
                            return false;
                        }
                        if (this._postData[key] !== this.malData[key]) {
                            return true;
                        };
                        return false;
                    }
                default:
                    {
                        // Treat falsy values as equivalent (!= doesn't do the trick here)
                        if (!this._postData[key] && !this.malData[key]) {
                            return false;
                        }
                        if (this._postData[key] !== this.malData[key]) {
                            return true;
                        }
                        return false;
                    }
            }
        });
    }

    shouldAdd(): boolean {
        return !this.malData;
    }

    formData(): string {
        throw new Error("Method not implemented.");
    }
    protected createPostData(): T.MALPostItem {
        throw new Error("Method not implemented.");
    }

    get type(): string {
        return this.alData.type;
    }
    get id(): number {
        return this.alData.id;
    }
    get title(): string {
        return this.alData.title;
    }
    get postData(): T.MALPostItem {
        return this._postData;
    }
}

export class MALEntryAnime extends BaseMALEntry {
    constructor(al: T.FormattedEntry, mal: T.MALLoadItem, csrfToken: string) {
        super(al, mal, csrfToken);
    }

    createPostData(): T.MALPostItem {
        const result = this.createBaseMALPostItem();
        result.anime_id = this.alData.id;

        if (this.alData.repeat) result.num_watched_times = this.alData.repeat;

        // If MAL episode count is available, use it as a maximum
        // For new items it will not be present; however the list will refresh after add and
        // it should be available then
        result.num_watched_episodes = this.malData && this.malData.anime_num_episodes ?
            Math.min(this.alData.progress, this.malData.anime_num_episodes) :
            this.alData.progress || 0;

        return result;
    }

    formData(): string {
        const formData = {
            anime_id: this.malData.anime_id,
            aeps: this._postData.anime_num_episodes || 0,
            astatus: this.malData.anime_airing_status,
            'add_anime[status]': this._postData.status,
            'add_anime[num_watched_episodes]': this._postData.num_watched_episodes || 0,
            'add_anime[score]': this._postData.score,
            'add_anime[start_date][month]': this._postData.start_date && this._postData.start_date.month || '',
            'add_anime[start_date][day]': this._postData.start_date && this._postData.start_date.day || '',
            'add_anime[start_date][year]': this._postData.start_date && this._postData.start_date.year || '',
            'add_anime[finish_date][month]': this._postData.finish_date && this._postData.finish_date.month || '',
            'add_anime[finish_date][day]': this._postData.finish_date && this._postData.finish_date.day || '',
            'add_anime[finish_date][year]': this._postData.finish_date && this._postData.finish_date.year || '',
            'add_anime[tags]': this.malData.tags || '',
            'add_anime[priority]': 0,
            'add_anime[storage_type]': '',
            'add_anime[storage_value]': 0,
            'add_anime[num_watched_times]': this._postData.num_watched_times || 0,
            'add_anime[rewatch_value]': '',
            'add_anime[comments]': '',
            'add_anime[is_asked_to_discuss]': 0,
            'add_anime[sns_post_type]': 0,
            submitIt: 0,
            csrf_token: this.csrfToken,
        } as T.MALFormData;
        if (this.alData.status === 'REPEATING') {
            formData['add_anime[is_rewatching]'] = 1;
        }
        return createMALFormData(formData);
    }
}

export class MALEntryManga extends BaseMALEntry {
    constructor(al: T.FormattedEntry, mal: T.MALLoadItem, csrfToken: string) {
        super(al, mal, csrfToken);
    }

    createPostData(): T.MALPostItem {
        const result = this.createBaseMALPostItem();
        result.manga_id = this.alData.id;

        if (this.alData.repeat) result.num_read_times = this.alData.repeat;

        // If MAL chapter and volume counts are available, use them as a maximum
        // For new items they will not be present; however the list will refresh after add and
        // they should be available then
        result.num_read_chapters = this.malData && this.malData.manga_num_chapters ?
            Math.min(this.alData.progress, this.malData.manga_num_chapters) :
            this.alData.progress || 0;

        result.num_read_volumes = this.malData && this.malData.manga_num_volumes ?
            Math.min(this.alData.progressVolumes, this.malData.manga_num_volumes) :
            this.alData.progressVolumes || 0;

        return result;
    }

    formData(): string {
        const formData = {
            entry_id: 0,
            manga_id: this.malData.manga_id,
            'add_manga[status]': this._postData.status,
            'add_manga[num_read_volumes]': this._postData.num_read_volumes || 0,
            last_completed_vol: '',
            'add_manga[num_read_chapters]': this._postData.num_read_chapters || 0,
            'add_manga[score]': this._postData.score || '',
            'add_manga[start_date][month]': this._postData.start_date && this._postData.start_date.month || '',
            'add_manga[start_date][day]': this._postData.start_date && this._postData.start_date.day || '',
            'add_manga[start_date][year]': this._postData.start_date && this._postData.start_date.year || '',
            'add_manga[finish_date][month]': this._postData.finish_date && this._postData.finish_date.month || '',
            'add_manga[finish_date][day]': this._postData.finish_date && this._postData.finish_date.day || '',
            'add_manga[finish_date][year]': this._postData.finish_date && this._postData.finish_date.year || '',
            'add_manga[tags]': this.malData.tags || '',
            'add_manga[priority]': 0,
            'add_manga[storage_type]': '',
            'add_manga[num_retail_volumes]': this.malData.manga_num_volumes || 0,
            'add_manga[num_read_times]': this._postData.num_read_times || 0,
            'add_manga[reread_value]': '',
            'add_manga[comments]': '',
            'add_manga[is_asked_to_discuss]': 0,
            'add_manga[sns_post_type]': 0,
            csrf_token: this.csrfToken,
            submitIt: 0
        } as T.MALFormData;
        if (this.alData.status === 'REPEATING') {
            formData['add_manga[is_rewatching]'] = 1;
        }
        return createMALFormData(formData);
    }
}