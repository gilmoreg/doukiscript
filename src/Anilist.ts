import * as Log from './Log';

/*
Anilist response is as follows:
data: {
    anime: {
        lists: [
            {
                entries: [
                    {entry}
                ]
            }
        ]
    },
    manga: {
        lists: [
            {
                entries: [
                    {entry}
                ]
            }
        ]
    },
}
*/

type AnilistDate = {
    year: number
    month: number
    day: number
}

type BaseEntry = {
    status: string
    score: number
    progress: number
    progressVolumes: number
    startedAt: AnilistDate
    completedAt: AnilistDate
    repeat: number
}

type Entry = BaseEntry & {
    media: {
        idMal: number
        title: {
            romaji: string
        }
    }
}

type MediaList = {
    entries: Array<Entry>
    [key: string]: Array<Entry>
}

type MediaListCollection = {
    lists: MediaList
}

type AnilistResponse = {
    anime: MediaListCollection
    manga: MediaListCollection
    [key: string]: MediaListCollection
}

type FormattedEntry = BaseEntry & {
    type: string
    id: number
    title: string
}

type DoukiAnilistData = {
    anime: Array<FormattedEntry>
    manga: Array<FormattedEntry>
}

const flatten = (obj: MediaList) =>
    // Outer reduce concats arrays built by inner reduce
    Object.keys(obj).reduce((accumulator, list) =>
        // Inner reduce builds an array out of the lists
        accumulator.concat(Object.keys(obj[list]).reduce((acc2, item) =>
            // @ts-ignore
            acc2.concat(obj[list][item]), [])), []);

const uniqify = (arr: Array<Entry>) => {
    const seen = new Set();
    return arr.filter(item => (seen.has(item.media.idMal) ? false : seen.add(item.media.idMal)));
};

// Anilist Functions
const anilistCall = (query: string, variables: any): Promise<Response> =>
    fetch('https://graphql.anilist.co', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            Accept: 'application/json',
        },
        body: JSON.stringify({
            query,
            variables,
        }),
    });

const fetchList = (userName: string) =>
    anilistCall(`
      query ($userName: String) {
        anime: MediaListCollection(userName: $userName, type: ANIME) {
          lists {
            entries {
              status
              score(format:POINT_10)
              progress
              startedAt {
                year
                month
                day
              }
              completedAt {
                year
                month
                day
              }
              repeat
              media {
                idMal
                title {
                  romaji
                }
              }
            }
          }
        },
        manga: MediaListCollection(userName: $userName, type: MANGA) {
          lists {
            entries {
              status
              score(format:POINT_10)
              progress
              progressVolumes
              startedAt {
                year
                month
                day
              }
              completedAt {
                year
                month
                day
              }
              repeat
              media {
                idMal
                title {
                  romaji
                }
              }
            }
          }
        }
      }
    `, {
            userName
        })
        .then(res => res.json())
        .then(res => res.data)
        .then(res => ({
            anime: uniqify(flatten(res.anime.lists)),
            manga: uniqify(flatten(res.manga.lists)),
        }));

const sanitize = (item: Entry, type: string): FormattedEntry => ({
    type,
    progress: item.progress,
    progressVolumes: item.progressVolumes,
    startedAt: {
        year: item.startedAt.year || 0,
        month: item.startedAt.month || 0,
        day: item.startedAt.day || 0,
    },
    completedAt: {
        year: item.completedAt.year || 0,
        month: item.completedAt.month || 0,
        day: item.completedAt.day || 0
    },
    repeat: item.repeat,
    status: item.status,
    score: item.score,
    id: item.media.idMal,
    title: item.media.title.romaji
});

const filterNoMalId = (item: FormattedEntry) => {
    if (item.id) return true;
    Log.error(`${item.type}: ${item.title}`);
    return false;
}

export const getAnilistList = (username: string): Promise<DoukiAnilistData> =>
    fetchList(username)
        .then(lists => ({
            anime: lists.anime
                .map(item => sanitize(item, 'anime'))
                .filter(item => filterNoMalId(item)),
            manga: lists.manga
                .map(item => sanitize(item, 'manga'))
                .filter(item => filterNoMalId(item)),
        }));