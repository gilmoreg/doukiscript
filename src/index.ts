import Log from './Log';
import Dom from './Dom';
import { getAnilistList } from './Anilist';
import MAL from './MAL';

// Main business logic
const sync = async (e: Event) => {
    e.preventDefault();

    const anilistUsername = Dom.getAnilistUsername();
    if (!anilistUsername) return;

    const malUsername = Dom.getMALUsername();
    if (!malUsername) {
        Log.info('You must be logged in!');
        return;
    }

    const csrfToken = Dom.getCSRFToken();

    Log.clear();
    Log.info(`Fetching data from Anilist...`);

    const anilistList = await getAnilistList(anilistUsername);
    if (!anilistList) {
        Log.info(`No data found for user ${anilistUsername}.`);
        return;
    }
    Log.info(`Fetched Anilist data.`);

    const mal = new MAL(malUsername, csrfToken);
    await mal.syncType('anime', anilistList.anime);
    await mal.syncType('manga', anilistList.manga);
    Log.info('Import complete.');
};

// Entrypoint
(() => {
    'use strict';
    Dom.addDropDownItem();

    if (window.location.pathname === '/import.php') {
        Dom.addImportForm(sync);
    }
})();
