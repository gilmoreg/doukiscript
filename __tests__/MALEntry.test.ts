import { MALEntryAnime, MALEntryManga } from '../src/MALEntry';
import * as fakes from '../__testutils__/testData';
import { AssertionError } from 'assert';

const fakeDomMethods = fakes.createFakeDomMethods();

const parseFormData = (formData: string) => {
    const result: any = {};
    formData.split('&').forEach(i => {
        const [key, value] = i.split('=');
        result[key] = value;
    });
    return result;
}

describe('shouldUpdate()', () => {
    it('should not update if data is the same', () => {
        const malAnime = fakes.createFakeMALAnime();
        const alAnime = fakes.createFakeAnilistAnime();
        const malEntry = new MALEntryAnime(alAnime, malAnime, 'csrfToken', fakeDomMethods);
        const result = malEntry.shouldUpdate();
        expect(result).toEqual(false);
    });

    it('should update if the start date is different', () => {
        const malAnime = fakes.createFakeMALAnime({ start_date_string: '2-2-2002' });
        const alAnime = fakes.createFakeAnilistAnime({ startedAt: { year: 2002, month: 1, day: 1 } });
        const malEntry = new MALEntryAnime(alAnime, malAnime, 'csrfToken', fakeDomMethods);
        const result = malEntry.shouldUpdate();
        expect(result).toEqual(true);
    });

    it('should update if the episode count is different and show is incomplete', () => {
        const malAnime = fakes.createFakeMALAnime({ status: 1, num_watched_episodes: 1 });
        const alAnime = fakes.createFakeAnilistAnime({ status: 'CURRENT', proress: 2 });
        const malEntry = new MALEntryAnime(alAnime, malAnime, 'csrfToken', fakeDomMethods);
        const result = malEntry.shouldUpdate();
        expect(result).toEqual(true);
    });

    it('should not update if episode counts differ but show is complete', () => {
        const malAnime = fakes.createFakeMALAnime({ num_watched_episodes: 1 });
        const alAnime = fakes.createFakeAnilistAnime({ progress: 2 });
        const malEntry = new MALEntryAnime(alAnime, malAnime, 'csrfToken', fakeDomMethods);
        const result = malEntry.shouldUpdate();
        expect(result).toEqual(false);
    });
});

describe('formData()', () => {
    it('should return default values of correct types for anime', async () => {
        const malAnime = fakes.createFakeMALAnime();
        const alAnime = fakes.createFakeAnilistAnime();
        const malEntry = new MALEntryAnime(alAnime, malAnime, 'csrfToken', fakeDomMethods);
        const result = await malEntry.formData();
        const json = parseFormData(result);
        expect(json).toMatchSnapshot();
        expect(json['add_anime%5Bis_rewatching%5D']).toBeUndefined();
    });

    it('should return default values of correct types for manga', async () => {
        const malManga = fakes.createFakeMALManga();
        const alManga = fakes.createFakeAnilistManga();
        const malEntry = new MALEntryManga(alManga, malManga, 'csrfToken', fakeDomMethods);
        const result = await malEntry.formData();
        const json = parseFormData(result);
        expect(json).toMatchSnapshot();
        expect(json['add_manga%5Bis_rewatching%5D']).toBeUndefined();
    });

    it('should set rewatching for anime', async () => {
        const malAnime = fakes.createFakeMALAnime();
        const alAnime = fakes.createFakeAnilistAnime({ status: 'REPEATING' });
        const malEntry = new MALEntryAnime(alAnime, malAnime, 'csrfToken', fakeDomMethods);
        const result = await malEntry.formData();
        const json = parseFormData(result);
        expect(json['add_anime%5Bis_rewatching%5D']).toEqual('1');
    });

    it('should set rewatching for manga', async () => {
        const malManga = fakes.createFakeMALManga();
        const alManga = fakes.createFakeAnilistManga({ status: 'REPEATING' });;
        const malEntry = new MALEntryManga(alManga, malManga, 'csrfToken', fakeDomMethods);
        const result = await malEntry.formData();
        const json = parseFormData(result);
        expect(json['add_manga%5Bis_rewatching%5D']).toEqual('1');
    });
});

describe('createPostData()', () => {
    it('should use AL data if MAL data is not available for anime', () => {
        const alAnime = fakes.createFakeAnilistAnime();
        // @ts-ignore
        const malEntry = new MALEntryAnime(alAnime, undefined, 'csrfToken', fakeDomMethods);
        expect(malEntry._postData.num_watched_episodes).toEqual(alAnime.progress);
    });

    it('should use AL data if MAL data is not available for manga', () => {
        const alManga = fakes.createFakeAnilistAnime();
        // @ts-ignore
        const malEntry = new MALEntryManga(alManga, undefined, 'csrfToken', fakeDomMethods);
        expect(malEntry._postData.num_read_chapters).toEqual(alManga.progress);
        expect(malEntry._postData.num_read_volumes).toEqual(alManga.progressVolumes);
    });

    it('uses MAL count if AL count is higher for completed manga', () => {
        const malManga = fakes.createFakeMALManga({ manga_num_chapters: 11, manga_num_volumes: 1 });
        const alManga = fakes.createFakeAnilistManga({ progress: 12, progressVolumes: 2 });
        const malEntry = new MALEntryManga(alManga, malManga, 'csrfToken', fakeDomMethods);
        expect(malEntry._postData.num_read_chapters).toEqual(11);
        expect(malEntry._postData.num_read_volumes).toEqual(1);
    });
});