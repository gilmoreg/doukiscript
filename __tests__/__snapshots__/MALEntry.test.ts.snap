// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formData() should return default values of correct types for anime 1`] = `
Object {
  "add_anime%5Bcomments%5D": "comments",
  "add_anime%5Bfinish_date%5D%5Bday%5D": "1",
  "add_anime%5Bfinish_date%5D%5Bmonth%5D": "1",
  "add_anime%5Bfinish_date%5D%5Byear%5D": "99",
  "add_anime%5Bis_asked_to_discuss%5D": "0",
  "add_anime%5Bnum_watched_episodes%5D": "12",
  "add_anime%5Bnum_watched_times%5D": "1",
  "add_anime%5Bpriority%5D": "0",
  "add_anime%5Brewatch_value%5D": "0",
  "add_anime%5Bscore%5D": "10",
  "add_anime%5Bsns_post_type%5D": "0",
  "add_anime%5Bstart_date%5D%5Bday%5D": "1",
  "add_anime%5Bstart_date%5D%5Bmonth%5D": "1",
  "add_anime%5Bstart_date%5D%5Byear%5D": "99",
  "add_anime%5Bstatus%5D": "2",
  "add_anime%5Bstorage_type%5D": "0",
  "add_anime%5Bstorage_value%5D": "0",
  "add_anime%5Btags%5D": "",
  "aeps": "12",
  "anime_id": "1",
  "astatus": "2",
  "csrf_token": "csrfToken",
  "g-recaptcha-response": "captcha",
  "submitIt": "0",
}
`;

exports[`formData() should return default values of correct types for manga 1`] = `
Object {
  "add_manga%5Bcomments%5D": "comments",
  "add_manga%5Bfinish_date%5D%5Bday%5D": "1",
  "add_manga%5Bfinish_date%5D%5Bmonth%5D": "1",
  "add_manga%5Bfinish_date%5D%5Byear%5D": "99",
  "add_manga%5Bis_asked_to_discuss%5D": "0",
  "add_manga%5Bnum_read_chapters%5D": "12",
  "add_manga%5Bnum_read_times%5D": "1",
  "add_manga%5Bnum_read_volumes%5D": "2",
  "add_manga%5Bnum_retail_volumes%5D": "0",
  "add_manga%5Bpriority%5D": "0",
  "add_manga%5Breread_value%5D": "0",
  "add_manga%5Bscore%5D": "10",
  "add_manga%5Bsns_post_type%5D": "0",
  "add_manga%5Bstart_date%5D%5Bday%5D": "1",
  "add_manga%5Bstart_date%5D%5Bmonth%5D": "1",
  "add_manga%5Bstart_date%5D%5Byear%5D": "99",
  "add_manga%5Bstatus%5D": "2",
  "add_manga%5Bstorage_type%5D": "0",
  "add_manga%5Btags%5D": "",
  "csrf_token": "csrfToken",
  "entry_id": "0",
  "g-recaptcha-response": "captcha",
  "last_completed_vol": "",
  "manga_id": "2",
  "submitIt": "0",
}
`;
